sharding/contracts/validator_manager.sol

```diff
@@ -0,0 +1,229 @@
+    emptySlotsStack[emptySlotsStackTop] = index;
+    ++emptySlotsStackTop;
+  }
+  function stackPop() internal returns(int) {
+    if (isStackEmpty())
+      return -1;
+    --emptySlotsStackTop;
+    return emptySlotsStack[emptySlotsStackTop];
+  }
+
+  function getValidatorsMaxIndex() internal view returns(int) {
+    int activateValidatorNum = 0;
+    int allValidatorSlotsNum = numValidators + emptySlotsStackTop;
+
+    // TODO: any better way to iterate the mapping?
+    for (int i = 0; i < 1024; ++i) {
+        if (i >= allValidatorSlotsNum)
+            break;
+        if (validators[i].addr != 0x0)
+            activateValidatorNum += 1;
+    }
+    return activateValidatorNum + emptySlotsStackTop;
+  }
+
+  function deposit() public payable returns(int) {
+    require(!isValidatorDeposited[msg.sender]);
+    require(msg.value == depositSize);
+    // Find the empty slot index in validators set
+    int index;
+    if (!isStackEmpty())
+      index = stackPop();
+    else
+      index = int(numValidators);
+      
+    validators[index] = Validator({
+      deposit: msg.value,
+      addr: msg.sender
+    });
+  ++numValidators;
+  isValidatorDeposited[msg.sender] = true;
+  
+  Deposit(msg.sender, index);
+  return index;
+  }
+
+  function withdraw(int _validatorIndex) public {
+    require(msg.sender == validators[_validatorIndex].addr);
+    // [FIXME] Should consider calling the validator's contract, might be useful
+    // when the validator is a contract.
+    validators[_validatorIndex].addr.transfer(validators[_validatorIndex].deposit);
+    isValidatorDeposited[validators[_validatorIndex].addr] = false;
+    delete validators[_validatorIndex];
+    stackPush(_validatorIndex);
+    --numValidators;
+    Withdraw(_validatorIndex);
+  }
+
+  // Uses a block hash as a seed to pseudorandomly select a signer from the validator set.
+  // [TODO] Chance of being selected should be proportional to the validator's deposit.
+  // Should be able to return a value for the current period or any future period up to.
+  function getEligibleProposer(int _shardId, uint _period) public view returns(address) {
+    require(_period >= lookAheadPeriods);
+    require((_period - lookAheadPeriods) * periodLength < block.number);
+    require(numValidators > 0);
+    // [TODO] Should check further if this safe or not
+    return validators[
+      int(
+      uint(keccak256(uint(block.blockhash(_period - lookAheadPeriods)) * periodLength, _shardId))
+      %
+      uint(getValidatorsMaxIndex())
+      )].addr;
+  }
+
+  struct HeaderVars {
+    bytes32 entireHeaderHash;
+    int score;
+    address validatorAddr;
+    bool isNewHead;
+  }
+  function addHeader(int _shardId, uint _expectedPeriodNumber, bytes32 _periodStartPrevHash,
+                     bytes32 _parentCollationHash, bytes32 _txListRoot, address _collationCoinbase,
+                     bytes32 _postStateRoot, bytes32 _receiptRoot, int _collationNumber) public returns(bool) {
```

Comment:
Always returns true? When is a false state? Can this be a void method?

Reply:
I think it can be a void method. We should probably clone also the spec at https://github.com/ethereum/sharding/blob/develop/docs/doc.md if we start differing so much, no?

Reply:
Sounds good to me